File: windows/src/HoverPocket.Shell/Capabilities/ProviderCapabilityHandlers.cs
using System.Globalization;
using System.Text.Json;
using HoverPocket.Shell.Providers.Calendar;
using HoverPocket.Shell.Providers.Sticky;
using HoverPocket.Shell.Providers.Timer;

namespace HoverPocket.Shell.Capabilities;

internal static class CapabilityIds
{
    public static readonly PocketCapabilityKey CalendarList = new("calendar.events.list", 1);
    public static readonly PocketCapabilityKey CalendarGet = new("calendar.event.get", 1);
    public static readonly PocketCapabilityKey CalendarCreate = new("calendar.event.create", 1);
    public static readonly PocketCapabilityKey TimerStart = new("timer.countdown.start", 1);
    public static readonly PocketCapabilityKey TimerGet = new("timer.countdown.get", 1);
    public static readonly PocketCapabilityKey TimerPause = new("timer.countdown.pause", 1);
    public static readonly PocketCapabilityKey TimerResume = new("timer.countdown.resume", 1);
    public static readonly PocketCapabilityKey TimerStop = new("timer.countdown.stop", 1);
    public static readonly PocketCapabilityKey StickyUpsert = new("sticky.note.upsert", 1);
    public static readonly PocketCapabilityKey StickyGet = new("sticky.note.get", 1);
}

internal sealed record CalendarCapabilityEvent(
    string EventRef,
    string EventId,
    string SafeTitle,
    DateTimeOffset Start,
    DateTimeOffset End);

internal sealed record CalendarCapabilityCreateRequest(
    string? CalendarId,
    string Title,
    DateTimeOffset Start,
    DateTimeOffset End,
    bool IsAllDay,
    string? Location,
    string? Notes);

internal interface ICalendarCapabilityDataSource
{
    Task<IReadOnlyList<CalendarCapabilityEvent>> ListEventsAsync(
        DateTimeOffset start,
        DateTimeOffset end,
        CancellationToken cancellationToken);

    Task<CalendarCapabilityEvent?> GetEventAsync(
        string eventRef,
        CancellationToken cancellationToken);

    Task<CalendarCapabilityEvent> CreateEventAsync(
        CalendarCapabilityCreateRequest request,
        string idempotencyKey,
        CancellationToken cancellationToken);
}

internal sealed class GoogleCalendarCapabilityDataSource : ICalendarCapabilityDataSource
{
    private readonly CalendarStore _store;

    public GoogleCalendarCapabilityDataSource(CalendarStore store)
    {
        _store = store;
    }

    public async Task<IReadOnlyList<CalendarCapabilityEvent>> ListEventsAsync(
        DateTimeOffset start,
        DateTimeOffset end,
        CancellationToken cancellationToken)
    {
        var events = await _store.ListEventsForCapabilityAsync(start, end, cancellationToken);
        return events.Select(Map).ToArray();
    }

    public async Task<CalendarCapabilityEvent?> GetEventAsync(
        string eventRef,
        CancellationToken cancellationToken)
    {
        var item = await _store.GetEventForCapabilityAsync(eventRef, cancellationToken);
        return item is null ? null : Map(item);
    }

    public async Task<CalendarCapabilityEvent> CreateEventAsync(
        CalendarCapabilityCreateRequest request,
        string idempotencyKey,
        CancellationToken cancellationToken)
    {
        return Map(await _store.CreateEventForCapabilityAsync(
            request,
            idempotencyKey,
            cancellationToken));
    }

    private static CalendarCapabilityEvent Map(CalendarEventOccurrence item)
    {
        return new CalendarCapabilityEvent(
            item.Id,
            item.GoogleEventId,
            item.Title,
            item.Start,
            item.End);
    }
}

internal sealed class CalendarListCapabilityHandler : IPocketCapabilityHandler
{
    private readonly ICalendarCapabilityDataSource _dataSource;

    public CalendarListCapabilityHandler(ICalendarCapabilityDataSource dataSource)
    {
        _dataSource = dataSource;
    }

    public PocketCapabilityKey Key => CapabilityIds.CalendarList;

    public async Task<JsonElement> HandleAsync(
        JsonElement arguments,
        CapabilityHandlerContext context,
        CancellationToken cancellationToken = default)
    {
        if (CapabilityJson.RequiredString(arguments, "range", 16) != "today")
        {
            throw CapabilityJson.Invalid("range");
        }
        var timeZoneId = CapabilityJson.RequiredString(arguments, "timezone", 64);
        TimeZoneInfo timeZone;
        try
        {
            timeZone = TimeZoneInfo.FindSystemTimeZoneById(timeZoneId);
        }
        catch (TimeZoneNotFoundException)
        {
            throw CapabilityJson.Invalid("timezone");
        }
        catch (InvalidTimeZoneException)
        {
            throw CapabilityJson.Invalid("timezone");
        }

        var localNow = TimeZoneInfo.ConvertTime(context.Now, timeZone);
        var localDate = DateOnly.FromDateTime(localNow.DateTime);
        var localStart = StartOfDay(localDate, timeZone);
        var localEnd = StartOfDay(localDate.AddDays(1), timeZone);
        var events = await _dataSource.ListEventsAsync(localStart, localEnd, cancellationToken);
        return CapabilityJson.From(new
        {
            events = events.Take(128).Select(Output).ToArray()
        });
    }

    internal static object Output(CalendarCapabilityEvent item)
    {
        return new
        {
            EventRef = CapabilityJson.OutputString(item.EventRef, 256, "calendar.eventRef"),
            start = item.Start.ToString("O", CultureInfo.InvariantCulture),
            end = item.End.ToString("O", CultureInfo.InvariantCulture),
            safeTitle = CapabilityJson.TruncateString(item.SafeTitle, 160)
        };
    }

    private static DateTimeOffset StartOfDay(DateOnly date, TimeZoneInfo timeZone)
    {
        var local = DateTime.SpecifyKind(date.ToDateTime(TimeOnly.MinValue), DateTimeKind.Unspecified);
        return new DateTimeOffset(local, timeZone.GetUtcOffset(local));
    }
}

internal sealed class CalendarGetCapabilityHandler : IPocketCapabilityHandler
{
    private readonly ICalendarCapabilityDataSource _dataSource;

    public CalendarGetCapabilityHandler(ICalendarCapabilityDataSource dataSource)
    {
        _dataSource = dataSource;
    }

    public PocketCapabilityKey Key => CapabilityIds.CalendarGet;

    public async Task<JsonElement> HandleAsync(
        JsonElement arguments,
        CapabilityHandlerContext context,
        CancellationToken cancellationToken = default)
    {
        _ = context;
        var eventRef = CapabilityJson.RequiredString(arguments, "eventRef", 256);
        var item = await _dataSource.GetEventAsync(eventRef, cancellationToken)
            ?? throw new CapabilityHandlerException("CAPABILITY_UNAVAILABLE", "calendar_event");
        return CapabilityJson.From(new
        {
            EventRef = CapabilityJson.OutputString(item.EventRef, 256, "calendar.eventRef"),
            EventId = CapabilityJson.OutputString(item.EventId, 256, "calendar.eventId"),
            start = item.Start.ToString("O", CultureInfo.InvariantCulture),
            end = item.End.ToString("O", CultureInfo.InvariantCulture),
            safeTitle = CapabilityJson.TruncateString(item.SafeTitle, 160)
        });
    }
}

internal sealed class CalendarCreateCapabilityHandler : IPocketCapabilityHandler
{
    private readonly ICalendarCapabilityDataSource _dataSource;

    public CalendarCreateCapabilityHandler(ICalendarCapabilityDataSource dataSource)
    {
        _dataSource = dataSource;
    }

    public PocketCapabilityKey Key => CapabilityIds.CalendarCreate;

    public async Task<JsonElement> HandleAsync(
        JsonElement arguments,
        CapabilityHandlerContext context,
        CancellationToken cancellationToken = default)
    {
        var idempotencyKey = context.RequireIdempotencyKey();
        var title = CapabilityJson.RequiredString(arguments, "title", 160);
        var start = RequiredDate(arguments, "start");
        var end = RequiredDate(arguments, "end");
        var isAllDay = CapabilityJson.RequiredBool(arguments, "isAllDay");
        if (isAllDay)
        {
            var startDate = DateOnly.FromDateTime(start.Date);
            var endDate = DateOnly.FromDateTime(end.Date);
            if (endDate <= startDate)
            {
                throw CapabilityJson.Invalid("start_end");
            }
        }
        else if (end <= start)
        {
            throw CapabilityJson.Invalid("start_end");
        }
        var request = new CalendarCapabilityCreateRequest(
            CapabilityJson.OptionalString(arguments, "calendarId", 256),
            title,
            start,
            end,
            isAllDay,
            CapabilityJson.OptionalString(arguments, "location", 500),
            CapabilityJson.OptionalString(arguments, "notes", 10_000));
        var created = await _dataSource.CreateEventAsync(
            request,
            idempotencyKey,
            cancellationToken);
        var observed = await _dataSource.GetEventAsync(created.EventRef, cancellationToken);
        if (observed != created)
        {
            throw new CapabilityHandlerException("CAPABILITY_READBACK_MISMATCH", "calendar.event.create");
        }
        return CapabilityJson.From(new
        {
            EventRef = CapabilityJson.OutputString(created.EventRef, 256, "calendar.eventRef"),
            EventId = CapabilityJson.OutputString(created.EventId, 256, "calendar.eventId"),
            start = created.Start.ToString("O", CultureInfo.InvariantCulture),
            end = created.End.ToString("O", CultureInfo.InvariantCulture),
            safeTitle = CapabilityJson.TruncateString(created.SafeTitle, 160)
        });
    }

    private static DateTimeOffset RequiredDate(JsonElement arguments, string name)
    {
        var value = CapabilityJson.RequiredString(arguments, name, 64);
        if (!DateTimeOffset.TryParse(value, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out var parsed))
        {
            throw CapabilityJson.Invalid(name);
        }
        return parsed;
    }
}

internal enum TimerCapabilityOperation
{
    Start,
    Get,
    Pause,
    Resume,
    Stop
}

internal sealed class TimerCapabilityHandler : IPocketCapabilityHandler
{
    private readonly TimerCapabilityOperation _operation;
    private readonly TimerStore _store;

    public TimerCapabilityHandler(TimerCapabilityOperation operation, TimerStore store)
    {
        _operation = operation;
        _store = store;
    }

    public PocketCapabilityKey Key => _operation switch
    {
        TimerCapabilityOperation.Start => CapabilityIds.TimerStart,
        TimerCapabilityOperation.Get => CapabilityIds.TimerGet,
        TimerCapabilityOperation.Pause => CapabilityIds.TimerPause,
        TimerCapabilityOperation.Resume => CapabilityIds.TimerResume,
        TimerCapabilityOperation.Stop => CapabilityIds.TimerStop,
        _ => throw new ArgumentOutOfRangeException()
    };

    public Task<JsonElement> HandleAsync(
        JsonElement arguments,
        CapabilityHandlerContext context,
        CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();
        if (_operation != TimerCapabilityOperation.Get)
        {
            _ = context.RequireIdempotencyKey();
        }
        try
        {
            return Task.FromResult(_operation switch
            {
                TimerCapabilityOperation.Start => Start(arguments),
                TimerCapabilityOperation.Get => Output(TimerId(arguments)),
                TimerCapabilityOperation.Pause => Pause(TimerId(arguments)),
                TimerCapabilityOperation.Resume => Resume(TimerId(arguments)),
                TimerCapabilityOperation.Stop => Stop(TimerId(arguments)),
                _ => throw new ArgumentOutOfRangeException()
            });
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            throw new CapabilityHandlerException("CAPABILITY_UNAVAILABLE", "timer_storage");
        }
    }

    private JsonElement Start(JsonElement arguments)
    {
        var duration = CapabilityJson.RequiredInt(arguments, "durationSeconds", 1, 86_400);
        var title = CapabilityJson.RequiredString(arguments, "title", 80);
        _ = CapabilityJson.OptionalString(arguments, "sourceRef", 256);
        var before = _store.GetSnapshot().RunningTimers.Select(timer => timer.Id).ToHashSet();
        var preset = TimerPreset.DefaultTimerDraft() with
        {
            Title = title,
            DurationSeconds = duration
        };
        var snapshot = _store.Start(preset);
        var created = snapshot.RunningTimers.SingleOrDefault(timer => !before.Contains(timer.Id))
            ?? throw new CapabilityHandlerException("CAPABILITY_UNAVAILABLE", "timer_capacity");
        return Output(created);
    }

    private JsonElement Pause(Guid id)
    {
        if (_store.GetRunningTimer(id) is null)
        {
            throw new CapabilityHandlerException("CAPABILITY_UNAVAILABLE", "timer");
        }
        _store.Pause(id);
        return Output(id);
    }

    private JsonElement Resume(Guid id)
    {
        if (_store.GetRunningTimer(id) is null)
        {
            throw new CapabilityHandlerException("CAPABILITY_UNAVAILABLE", "timer");
        }
        _store.Resume(id);
        return Output(id);
    }

    private JsonElement Stop(Guid id)
    {
        _store.Stop(id);
        return Stopped(id);
    }

    private JsonElement Output(Guid id)
    {
        var timer = _store.GetSnapshot().RunningTimers.FirstOrDefault(item => item.Id == id);
        return timer is null ? Stopped(id) : Output(timer);
    }

    private static JsonElement Output(RunningTimerSnapshot timer)
    {
        return CapabilityJson.From(new
        {
            timerId = timer.Id.ToString("D").ToLowerInvariant(),
            state = timer.IsPaused ? "paused" : "running",
            endAt = timer.EndAtUtc.ToString("O", CultureInfo.InvariantCulture)
        });
    }

    private static JsonElement Stopped(Guid id)
    {
        return CapabilityJson.From(new
        {
            timerId = id.ToString("D").ToLowerInvariant(),
            state = "stopped",
            endAt = (string?)null
        });
    }

    private static Guid TimerId(JsonElement arguments)
    {
        var value = CapabilityJson.RequiredString(arguments, "timerId", 36);
        return Guid.TryParse(value, out var id) ? id : throw CapabilityJson.Invalid("timerId");
    }
}

internal enum StickyCapabilityOperation
{
    Upsert,
    Get
}

internal sealed class StickyCapabilityHandler : IPocketCapabilityHandler
{
    private readonly StickyCapabilityOperation _operation;
    private readonly StickyNotesStore _store;

    public StickyCapabilityHandler(StickyCapabilityOperation operation, StickyNotesStore store)
    {
        _operation = operation;
        _store = store;
    }

    public PocketCapabilityKey Key => _operation == StickyCapabilityOperation.Upsert
        ? CapabilityIds.StickyUpsert
        : CapabilityIds.StickyGet;

    public Task<JsonElement> HandleAsync(
        JsonElement arguments,
        CapabilityHandlerContext context,
        CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();
        if (_operation == StickyCapabilityOperation.Upsert)
        {
            _ = context.RequireIdempotencyKey();
        }
        return Task.FromResult(_operation == StickyCapabilityOperation.Upsert
            ? Upsert(arguments)
            : Get(arguments));
    }

    private JsonElement Upsert(JsonElement arguments)
    {
        StickyNoteItem note;
        try
        {
            note = _store.UpsertNote(
                CapabilityJson.RequiredString(arguments, "stableKey", 160),
                CapabilityJson.RequiredString(arguments, "title", 120, allowEmpty: true),
                CapabilityJson.RequiredString(arguments, "body", 10_000, allowEmpty: true),
                Color(CapabilityJson.RequiredString(arguments, "color", 16)));
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            throw new CapabilityHandlerException("CAPABILITY_UNAVAILABLE", "sticky_storage");
        }
        return MutationOutput(note);
    }

    private JsonElement Get(JsonElement arguments)
    {
        var rawId = CapabilityJson.RequiredString(arguments, "noteId", 128);
        if (!Guid.TryParse(rawId, out var id) || _store.GetNote(id) is not { } note)
        {
            throw new CapabilityHandlerException("CAPABILITY_UNAVAILABLE", "sticky_note");
        }
        return CapabilityJson.From(new
        {
            noteId = note.Id.ToString("D").ToLowerInvariant(),
            Title = CapabilityJson.OutputString(note.Title, 120, "sticky.title", allowEmpty: true),
            Body = CapabilityJson.OutputString(note.Body, 10_000, "sticky.body", allowEmpty: true),
            updatedAt = note.UpdatedAt.ToString("O", CultureInfo.InvariantCulture)
        });
    }

    private static JsonElement MutationOutput(StickyNoteItem note)
    {
        return CapabilityJson.From(new
        {
            noteId = note.Id.ToString("D").ToLowerInvariant(),
            Title = CapabilityJson.OutputString(note.Title, 120, "sticky.title", allowEmpty: true),
            Body = CapabilityJson.OutputString(note.Body, 10_000, "sticky.body", allowEmpty: true),
            updatedAt = note.UpdatedAt.ToString("O", CultureInfo.InvariantCulture)
        });
    }

    private static StickyNoteColor Color(string value)
    {
        return value switch
        {
            "yellow" => StickyNoteColor.Yellow,
            "blue" => StickyNoteColor.Blue,
            "green" => StickyNoteColor.Mint,
            "pink" => StickyNoteColor.Pink,
            "gray" => StickyNoteColor.Lavender,
            _ => throw CapabilityJson.Invalid("color")
        };
    }
}

internal static class ProviderCapabilityCompositionRoot
{
    public static PocketCapabilityHandlerSet Create(
        ICalendarCapabilityDataSource calendarDataSource,
        TimerStore timerStore,
        StickyNotesStore stickyStore)
    {
        return new PocketCapabilityHandlerSet([
            new CalendarListCapabilityHandler(calendarDataSource),
            new CalendarGetCapabilityHandler(calendarDataSource),
            new CalendarCreateCapabilityHandler(calendarDataSource),
            new TimerCapabilityHandler(TimerCapabilityOperation.Start, timerStore),
            new TimerCapabilityHandler(TimerCapabilityOperation.Get, timerStore),
            new TimerCapabilityHandler(TimerCapabilityOperation.Pause, timerStore),
            new TimerCapabilityHandler(TimerCapabilityOperation.Resume, timerStore),
            new TimerCapabilityHandler(TimerCapabilityOperation.Stop, timerStore),
            new StickyCapabilityHandler(StickyCapabilityOperation.Upsert, stickyStore),
            new StickyCapabilityHandler(StickyCapabilityOperation.Get, stickyStore)
        ]);
    }
}

File: Sources/HoverPocket/Capabilities/ProviderCapabilityHandlers.swift
import Foundation

private enum CapabilityIDs {
    static let calendarList = PocketCapabilityKey(id: "calendar.events.list", version: 1)
    static let calendarGet = PocketCapabilityKey(id: "calendar.event.get", version: 1)
    static let calendarCreate = PocketCapabilityKey(id: "calendar.event.create", version: 1)
    static let timerStart = PocketCapabilityKey(id: "timer.countdown.start", version: 1)
    static let timerGet = PocketCapabilityKey(id: "timer.countdown.get", version: 1)
    static let timerPause = PocketCapabilityKey(id: "timer.countdown.pause", version: 1)
    static let timerResume = PocketCapabilityKey(id: "timer.countdown.resume", version: 1)
    static let timerStop = PocketCapabilityKey(id: "timer.countdown.stop", version: 1)
    static let stickyUpsert = PocketCapabilityKey(id: "sticky.note.upsert", version: 1)
    static let stickyGet = PocketCapabilityKey(id: "sticky.note.get", version: 1)
}

struct CalendarCapabilityEvent: Equatable, Sendable {
    let eventRef: String
    let eventID: String
    let safeTitle: String
    let start: Date
    let end: Date
}

struct CalendarCapabilityCreateRequest: Equatable, Sendable {
    let calendarID: String?
    let title: String
    let start: Date
    let end: Date
    let isAllDay: Bool
    let allDayStart: CalendarCivilDate?
    let allDayEnd: CalendarCivilDate?
    let location: String?
    let notes: String?
}

struct CalendarCivilDate: Equatable, Sendable, Comparable {
    let year: Int
    let month: Int
    let day: Int

    init?(rfc3339 value: String) {
        let components = value.prefix(10).split(separator: "-", omittingEmptySubsequences: false)
        guard components.count == 3,
              let year = Int(components[0]),
              let month = Int(components[1]),
              let day = Int(components[2]) else {
            return nil
        }
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        let candidate = DateComponents(
            calendar: calendar,
            timeZone: calendar.timeZone,
            year: year,
            month: month,
            day: day
        )
        guard let date = calendar.date(from: candidate),
              calendar.dateComponents([.year, .month, .day], from: date)
                == DateComponents(year: year, month: month, day: day) else {
            return nil
        }
        self.year = year
        self.month = month
        self.day = day
    }

    func date(in calendar: Calendar) -> Date? {
        calendar.date(from: DateComponents(
            calendar: calendar,
            timeZone: calendar.timeZone,
            year: year,
            month: month,
            day: day
        ))
    }

    static func < (lhs: CalendarCivilDate, rhs: CalendarCivilDate) -> Bool {
        if lhs.year != rhs.year { return lhs.year < rhs.year }
        if lhs.month != rhs.month { return lhs.month < rhs.month }
        return lhs.day < rhs.day
    }
}

@MainActor
protocol CalendarCapabilityDataSource: AnyObject {
    func listEvents(from start: Date, to end: Date) async throws -> [CalendarCapabilityEvent]
    func getEvent(eventRef: String) async throws -> CalendarCapabilityEvent?
    func createEvent(
        _ request: CalendarCapabilityCreateRequest,
        idempotencyKey: String
    ) async throws -> CalendarCapabilityEvent
}

@MainActor
final class GoogleCalendarCapabilityDataSource: CalendarCapabilityDataSource {
    private let store: GoogleCalendarStore

    init(store: GoogleCalendarStore = .shared) {
        self.store = store
    }

    func listEvents(from start: Date, to end: Date) async throws -> [CalendarCapabilityEvent] {
        try await store.listEventsForCapability(from: start, to: end).map(Self.map)
    }

    func getEvent(eventRef: String) async throws -> CalendarCapabilityEvent? {
        try await store.eventForCapability(eventRef: eventRef).map(Self.map)
    }

    func createEvent(
        _ request: CalendarCapabilityCreateRequest,
        idempotencyKey: String
    ) async throws -> CalendarCapabilityEvent {
        Self.map(try await store.createEventForCapability(request, idempotencyKey: idempotencyKey))
    }

    private static func map(_ event: GoogleCalendarEventOccurrence) -> CalendarCapabilityEvent {
        CalendarCapabilityEvent(
            eventRef: event.id,
            eventID: event.googleEventID,
            safeTitle: event.title,
            start: event.start,
            end: event.end
        )
    }
}

@MainActor
final class CalendarListCapabilityHandler: PocketCapabilityHandler {
    let key = CapabilityIDs.calendarList
    private let dataSource: any CalendarCapabilityDataSource

    init(dataSource: any CalendarCapabilityDataSource) {
        self.dataSource = dataSource
    }

    func handle(arguments: CapabilityObject, context: CapabilityHandlerContext) async throws -> CapabilityObject {
        let range = try arguments.requiredString("range", maxLength: 16)
        guard range == "today" else {
            throw CapabilityHandlerError.invalidArgument("range")
        }
        let timeZoneID = try arguments.requiredString("timezone", maxLength: 64)
        guard let timeZone = TimeZone(identifier: timeZoneID) else {
            throw CapabilityHandlerError.invalidArgument("timezone")
        }
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone
        let start = calendar.startOfDay(for: context.now)
        guard let end = calendar.date(byAdding: .day, value: 1, to: start) else {
            throw CapabilityHandlerError.unavailable("calendar_range")
        }
        let events = try await dataSource.listEvents(from: start, to: end)
            .prefix(128)
            .map(Self.output)
        return ["events": .array(events.map(CapabilityValue.object))]
    }

    fileprivate static func output(_ event: CalendarCapabilityEvent) -> CapabilityObject {
        [
            "eventRef": .string(event.eventRef),
            "start": .string(CapabilityDateCodec.string(from: event.start)),
            "end": .string(CapabilityDateCodec.string(from: event.end)),
            "safeTitle": .string(event.safeTitle.prefixingUnicodeScalars(160))
        ]
    }
}

@MainActor
final class CalendarGetCapabilityHandler: PocketCapabilityHandler {
    let key = CapabilityIDs.calendarGet
    private let dataSource: any CalendarCapabilityDataSource

    init(dataSource: any CalendarCapabilityDataSource) {
        self.dataSource = dataSource
    }

    func handle(arguments: CapabilityObject, context: CapabilityHandlerContext) async throws -> CapabilityObject {
        _ = context
        let eventRef = try arguments.requiredString("eventRef", maxLength: 256)
        guard let event = try await dataSource.getEvent(eventRef: eventRef) else {
            throw CapabilityHandlerError.unavailable("calendar_event")
        }
        var output = CalendarListCapabilityHandler.output(event)
        output["eventId"] = .string(event.eventID)
        return output
    }
}

@MainActor
final class CalendarCreateCapabilityHandler: PocketCapabilityHandler {
    let key = CapabilityIDs.calendarCreate
    private let dataSource: any CalendarCapabilityDataSource

    init(dataSource: any CalendarCapabilityDataSource) {
        self.dataSource = dataSource
    }

    func handle(arguments: CapabilityObject, context: CapabilityHandlerContext) async throws -> CapabilityObject {
        let idempotencyKey = try context.requiredIdempotencyKey()
        let title = try arguments.requiredString("title", maxLength: 160)
        let startString = try arguments.requiredString("start", maxLength: 64)
        let endString = try arguments.requiredString("end", maxLength: 64)
        let isAllDay = try arguments.requiredBool("isAllDay")
        guard let start = CapabilityDateCodec.date(from: startString),
              let end = CapabilityDateCodec.date(from: endString) else {
            throw CapabilityHandlerError.invalidArgument("start_end")
        }
        let allDayStart = isAllDay ? CalendarCivilDate(rfc3339: startString) : nil
        let allDayEnd = isAllDay ? CalendarCivilDate(rfc3339: endString) : nil
        if isAllDay {
            guard let allDayStart, let allDayEnd, allDayEnd > allDayStart else {
                throw CapabilityHandlerError.invalidArgument("start_end")
            }
        } else if end <= start {
            throw CapabilityHandlerError.invalidArgument("start_end")
        }
        let request = CalendarCapabilityCreateRequest(
            calendarID: try arguments.optionalString("calendarId", maxLength: 256),
            title: title,
            start: start,
            end: end,
            isAllDay: isAllDay,
            allDayStart: allDayStart,
            allDayEnd: allDayEnd,
            location: try arguments.optionalString("location", maxLength: 500),
            notes: try arguments.optionalString("notes", maxLength: 10_000)
        )
        let created = try await dataSource.createEvent(request, idempotencyKey: idempotencyKey)
        guard let observed = try await dataSource.getEvent(eventRef: created.eventRef), observed == created else {
            throw CapabilityHandlerError.readbackMismatch("calendar.event.create")
        }
        return Self.output(observed)
    }

    private static func output(_ event: CalendarCapabilityEvent) -> CapabilityObject {
        var output = CalendarListCapabilityHandler.output(event)
        output["eventId"] = .string(event.eventID)
        return output
    }
}

@MainActor
final class TimerCapabilityHandler: PocketCapabilityHandler {
    enum Operation {
        case start
        case get
        case pause
        case resume
        case stop

        var key: PocketCapabilityKey {
            switch self {
            case .start: CapabilityIDs.timerStart
            case .get: CapabilityIDs.timerGet
            case .pause: CapabilityIDs.timerPause
            case .resume: CapabilityIDs.timerResume
            case .stop: CapabilityIDs.timerStop
            }
        }
    }

    let operation: Operation
    private let store: TimerStore
    private let idGenerator: @MainActor () -> UUID

    var key: PocketCapabilityKey { operation.key }

    init(operation: Operation, store: TimerStore, idGenerator: @escaping @MainActor () -> UUID = UUID.init) {
        self.operation = operation
        self.store = store
        self.idGenerator = idGenerator
    }

    func handle(arguments: CapabilityObject, context: CapabilityHandlerContext) async throws -> CapabilityObject {
        if operation != .get {
            _ = try context.requiredIdempotencyKey()
        }
        switch operation {
        case .start:
            return try await start(arguments: arguments, context: context)
        case .get:
            return try output(timerID: timerID(arguments), now: context.now)
        case .pause:
            let id = try timerID(arguments)
            guard store.runningTimer(id: id) != nil else {
                throw CapabilityHandlerError.unavailable("timer")
            }
            do {
                try await store.pauseForCapability(id: id, at: context.now)
            } catch {
                throw CapabilityHandlerError.unavailable("timer_storage")
            }
            return try output(timerID: id, now: context.now)
        case .resume:
            let id = try timerID(arguments)
            guard store.runningTimer(id: id) != nil else {
                throw CapabilityHandlerError.unavailable("timer")
            }
            do {
                try await store.resumeForCapability(id: id, at: context.now)
            } catch {
                throw CapabilityHandlerError.unavailable("timer_storage")
            }
            return try output(timerID: id, now: context.now)
        case .stop:
            let id = try timerID(arguments)
            do {
                try await store.stopForCapability(id: id)
            } catch {
                throw CapabilityHandlerError.unavailable("timer_storage")
            }
            return Self.stoppedOutput(id)
        }
    }

    private func start(arguments: CapabilityObject, context: CapabilityHandlerContext) async throws -> CapabilityObject {
        let duration = try arguments.requiredInteger("durationSeconds", range: 1...86_400)
        let title = try arguments.requiredString("title", maxLength: 80)
        _ = try arguments.optionalString("sourceRef", maxLength: 256)
        var preset = TimerPreset.defaultTimerDraft()
        preset.title = title
        preset.duration = TimeInterval(duration)
        let timer: RunningTimer?
        do {
            timer = try await store.startForCapability(preset: preset, id: idGenerator(), at: context.now)
        } catch {
            throw CapabilityHandlerError.unavailable("timer_storage")
        }
        guard let timer else {
            throw CapabilityHandlerError.unavailable("timer_capacity")
        }
        return Self.output(timer, now: context.now)
    }

    private func timerID(_ arguments: CapabilityObject) throws -> UUID {
        let value = try arguments.requiredString("timerId", maxLength: 36)
        guard let id = UUID(uuidString: value) else {
            throw CapabilityHandlerError.invalidArgument("timerId")
        }
        return id
    }

    private func output(timerID: UUID, now: Date) throws -> CapabilityObject {
        guard let timer = store.runningTimer(id: timerID) else {
            return Self.stoppedOutput(timerID)
        }
        return Self.output(timer, now: now)
    }

    private static func output(_ timer: RunningTimer, now: Date) -> CapabilityObject {
        [
            "timerId": .string(timer.id.uuidString.lowercased()),
            "state": .string(timer.isPaused ? "paused" : "running"),
            "endAt": .string(CapabilityDateCodec.string(from: timer.endDate))
        ]
    }

    private static func stoppedOutput(_ id: UUID) -> CapabilityObject {
        [
            "timerId": .string(id.uuidString.lowercased()),
            "state": .string("stopped"),
            "endAt": .null
        ]
    }
}

@MainActor
final class StickyCapabilityHandler: PocketCapabilityHandler {
    enum Operation {
        case upsert
        case get

        var key: PocketCapabilityKey {
            switch self {
            case .upsert: CapabilityIDs.stickyUpsert
            case .get: CapabilityIDs.stickyGet
            }
        }
    }

    let operation: Operation
    private let store: StickyNotesStore
    private let idGenerator: @MainActor () -> UUID

    var key: PocketCapabilityKey { operation.key }

    init(operation: Operation, store: StickyNotesStore, idGenerator: @escaping @MainActor () -> UUID = UUID.init) {
        self.operation = operation
        self.store = store
        self.idGenerator = idGenerator
    }

    func handle(arguments: CapabilityObject, context: CapabilityHandlerContext) async throws -> CapabilityObject {
        switch operation {
        case .upsert:
            _ = try context.requiredIdempotencyKey()
            let stableKey = try arguments.requiredString("stableKey", maxLength: 160)
            let title = try arguments.requiredString("title", maxLength: 120, allowEmpty: true)
            let body = try arguments.requiredString("body", maxLength: 10_000, allowEmpty: true)
            let color = try Self.color(try arguments.requiredString("color", maxLength: 16))
            let note: StickyNoteItem
            do {
                note = try store.upsertNote(
                    stableKey: stableKey,
                    title: title,
                    body: body,
                    color: color,
                    id: idGenerator(),
                    at: context.now
                )
            } catch {
                throw CapabilityHandlerError.unavailable("sticky_storage")
            }
            return Self.readOutput(note)
        case .get:
            let rawID = try arguments.requiredString("noteId", maxLength: 128)
            guard let id = UUID(uuidString: rawID), let note = store.note(id: id) else {
                throw CapabilityHandlerError.unavailable("sticky_note")
            }
            return Self.readOutput(note)
        }
    }

    private static func color(_ value: String) throws -> StickyNoteColor {
        switch value {
        case "yellow": .yellow
        case "blue": .blue
        case "green": .mint
        case "pink": .pink
        case "gray": .lavender
        default: throw CapabilityHandlerError.invalidArgument("color")
        }
    }

    private static func mutationOutput(_ note: StickyNoteItem) -> CapabilityObject {
        [
            "noteId": .string(note.id.uuidString.lowercased()),
            "updatedAt": .string(CapabilityDateCodec.string(from: note.updatedAt))
        ]
    }

    private static func readOutput(_ note: StickyNoteItem) -> CapabilityObject {
        var output = mutationOutput(note)
        output["title"] = .string(note.title)
        output["body"] = .string(note.body)
        return output
    }
}

@MainActor
enum ProviderCapabilityCompositionRoot {
    static func live(calendarDataSource: any CalendarCapabilityDataSource) throws -> PocketCapabilityHandlerSet {
        try PocketCapabilityHandlerSet(handlers: [
            CalendarListCapabilityHandler(dataSource: calendarDataSource),
            CalendarGetCapabilityHandler(dataSource: calendarDataSource),
            CalendarCreateCapabilityHandler(dataSource: calendarDataSource),
            TimerCapabilityHandler(operation: .start, store: .shared),
            TimerCapabilityHandler(operation: .get, store: .shared),
            TimerCapabilityHandler(operation: .pause, store: .shared),
            TimerCapabilityHandler(operation: .resume, store: .shared),
            TimerCapabilityHandler(operation: .stop, store: .shared),
            StickyCapabilityHandler(operation: .upsert, store: .shared),
            StickyCapabilityHandler(operation: .get, store: .shared)
        ])
    }
}
